src/adapter/philips_hue/mod.rs

```diff
@@ -0,0 +1,273 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate serde_json;
+
+mod http;
+mod nupnp;
+mod api;
+
+use adapter::philips_hue::api::structs::*;
+use adapter::philips_hue::api::HueLight;
+use controller::Controller;
+use events::*;
+use iron::{ Request, Response, IronResult };
+use iron::headers::ContentType;
+use iron::status::Status;
+use iron::method::Method;
+use router::Router;
+use service::{ Service, ServiceAdapter, ServiceProperties };
+use std::time::Duration;
+use std::thread;
+use uuid::Uuid;
+use std::io::Read;
+
+
+pub struct PhilipsHueAdapter<T> {
+    name: String,
+    controller: T,
+    hubs: Vec<api::HueApi>,
+}
+
+impl<T: Controller> PhilipsHueAdapter<T> {
+    pub fn new(controller: T) -> Self {
+        debug!("Creating Philips Hue adapter");
+        PhilipsHueAdapter { name: "PhilipsHueAdapter".to_owned(),
+                       controller: controller,
+                       hubs: nupnp::query(),
+                     }
+    }
+}
+
+impl<T: Controller> ServiceAdapter for PhilipsHueAdapter<T> {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn start(&self) {
+        let mut id = 0;
+        let controller = self.controller.clone();
+        let hubs = self.hubs.clone();
+
+        thread::spawn(move || {
+            controller.send_event(
+                EventData::AdapterStart {
+                    name: "Philips Hue Service Adapter".to_owned()
+                }).unwrap();
+
+            debug!("nUPnP reported Philips Hue bridges: {:?}", hubs);
+            'pairing_check: for hub in hubs {
+                hub.is_available();
+                info!("Philips Hue bridge {} at {}", hub.id, hub.ip);
+
+                // Querying the settings fails when not paired with the bridge
+                if !hub.is_paired() {
+                    controller.send_event(
+                        EventData::AdapterInfo {
+                            info: format!("Push the pairing button on your Philips Hue bridge with ID {}",
+                            hub.id)
+                        }).unwrap();
```

Comment:
I like the idea of sending adapter-level notifications! We'll need to add a bit more details to identify the adapter in the long run though.